reactioncommerce/reaction
# Fix #3518: AddressBook to react

Resolves #3518   
Impact: **major**  
Type: **feature|performance**

## Issue
The Address review workflow had to be redesigned.

## Solution
Whenever the client enters address that can't be validated by Avalara, show client a warning and give a chance to correct the address.
Also did performance optimizations for preventing hitting the Avalara API many times and prevented re-rendering of the AddressBook, because of fake updates to the Cart.

## Testing

## Test without tax
1. Login as admin
1. But an item and go to checkout screen
1. Try adding multiple addresses, changing default shipping and billing address
1. Try editing already added address
1. Everything should work.

## Test addressBook in profile
1. Login and go to profile section.
1. Try repeating above steps.

## Test with Avalara
1. Enable Avalara
1. Enable address validation
1. Enable tax calculation.
1. Add an item to cart and begin to checkout
1. For address enter RC office address
1. Observe it takes you to the review screen.
1. Click on `Edit entered address`, it should take you back to the edit address screen.
1. Again save the address with same mistakes.
1. It should take you to the review screen again.
1. Select the Suggested Address(should be selected by default)
1. Click on `Continue with selected address`.
1. It should save that address, and tax should be calulated.
1. Again repeat the above process, but on the review screen select the `Entered Address`
1. It should save the address but tax won't be calculated.
1. After placing the order, login as admin(if you weren't).
1. Go to the Orders from Action Menu, click on the above order, you should see it labelled `Tax not collected`.
